src/scene/vcPOI.cpp

```diff
@@ -1331,69 +1332,512 @@ void vcPOI::AddFillPolygonToScene(vcState *pProgramState, vcRenderData *pRenderD
+
+    if (angle < prevClosestAngle)
+    {
+      prevClosestAngle = angle;
+      currentPoint = pointIndex;
+    }
+  }
+
+  perimeterPoints.PushBack(startPoint);
+  while (currentPoint != startPoint)
+  {
+    perimeterPoints.PushBack(currentPoint);
+
+    // Reverse the previous angle
+    double prevAngle = -atan2(pointPositions[currentPoint].y - pointPositions[prevPoint].y, pointPositions[currentPoint].x - pointPositions[prevPoint].x);
+
+    int closestPoint = -1;
+    double closestAngle = DBL_MAX;
+    const udDouble3 &pos = pointPositions[currentPoint];
+    for (const int &pointIndex : pointConnections[currentPoint])
+      if (pointIndex != prevPoint)
+      {
+        const udDouble3 &target = pointPositions[pointIndex];
+        double angle = -atan2(pos.y - target.y, pos.x - target.x);
+
+        // Ensure the angle is positive relative to the previous angle
+        if (angle < prevAngle)
+          angle += UD_2PI;
+
+        if (angle < closestAngle)
+        {
+          closestAngle = angle;
+          closestPoint = pointIndex;
+        }
+      }
+
+    prevPoint = currentPoint;
+    currentPoint = closestPoint;
+  }
+
+  // Finalize
+  for (const int &pointIndex : perimeterPoints)
+    pPerimeterPoints->PushBack(pointPositions[pointIndex]);
+  perimeterPoints.Deinit();
+
+  pointPositions.Deinit();
+  for (udChunkedArray<int> &connections : pointConnections)
+    connections.Deinit();
+  pointConnections.Deinit();
+}
+
+bool vcPOI_CalculateTriangles(udDouble3 *pPoints, int numPoints, udChunkedArray<udDouble3> *pTrianglePoints)
+{
+  const size_t pointInfoChunkSize = 512;
+  const size_t pointInfoLinePointsChunkSize = 512;
+
+  udDouble3 pivotPoint = pPoints[0];
+
+  udChunkedArray<udDouble3> pointInfoPositions;
+  udChunkedArray<udChunkedArray<int>> pointInfoLines;
+  pointInfoPositions.Init(pointInfoChunkSize);
+  pointInfoLines.Init(pointInfoChunkSize);
+
+  for (int i = 0; i < numPoints; ++i)
+  {
+    int prevIndex = (i == 0 ? numPoints : i) - 1;
+    if (udMag3(pPoints[i] - pPoints[prevIndex]) < UD_EPSILON)
+      continue;
+
+    pointInfoPositions.PushBack(pPoints[i]);
+    pointInfoLines.PushBack();
+    pointInfoLines[pointInfoLines.length - 1].Init(pointInfoLinePointsChunkSize);
+  }
+    
+  // Remove duplicate Points
+  while (true)
+  {
+    bool unnecessaryPointRemoved = false;
+    for (size_t i = 0; i < pointInfoPositions.length; ++i)
+    {
+      int prevIndex = (int)(i == 0 ? (pointInfoPositions.length - 1) : i - 1);
+      int nextIndex = (int)(i == (pointInfoPositions.length - 1) ? 0 : i + 1);
+
+      udDouble2 curPointPos2 = udDouble2::create(pointInfoPositions[i].x, pointInfoPositions[i].y);
+      udDouble2 nextPointPos2 = udDouble2::create(pointInfoPositions[nextIndex].x, pointInfoPositions[nextIndex].y);
+      udDouble2 prevPointPos2 = udDouble2::create(pointInfoPositions[prevIndex].x, pointInfoPositions[prevIndex].y);
+
+      if (udMag2(curPointPos2 - prevPointPos2) < UD_EPSILON
```

Comment:
Not sure how often this runs, but potential optimization could be done here:

`udMagSq2() < UD_EPSILON * UD_EPSILON`